dotnet: improve rid handling

corngood opened the pull request:
###### Description of changes

Some things pulled out of #190129

@IvarWithoutBones 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

corngood committed:
omnisharp-roslyn: set runtime flag instead of patching project

corngood committed:
buildDotnetModule: run tests on projectFile if testProjectFile is unset

corngood committed:
build-dotnet-module: add useAppHost parameter so it can be disabled

corngood committed:
build-dotnet-module: restore for current runtime by default

corngood committed:
build-dotnet: make passthru.packages a derivation

corngood committed:
dotnetCorePackages: move systemToDotnetRid out of sdk

corngood committed:
build-dotnet-module: strip --runtime flags without using rid

corngood commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -163,11 +158,10 @@ stdenvNoCC.mkDerivation (args // {
         flags =
           let
-            hasRid = flag: lib.any (v: v) (map (rid: lib.hasInfix rid flag) (lib.attrValues dotnet-sdk.runtimeIdentifierMap));
+            isRuntime = flag: lib.hasPrefix "--runtime" flag;
```
this seems to be equivalent given what's currently in nixpkgs, and avoids a dependency on dotnetCorePackages, or exposing the list of known rids.

IvarWithoutBones replied:
I used the solution of relying on `runtimeIdentifiersMap` because there are many ways to pass them to dotnet, I figured this would be a bit more reliable. You could also pass one using an msbuild style flag, or the `-r` prefix, etc.

In my opinion relying on `dotnetCorePackages` is not a big deal since we don't propagate any of its outputs.

corngood replied:
Without thinking about it, I wrote a derivation that used `[ "--runtime" "linux-x64" ]` (which IMO is more natural than a single string), and it worked fine except for fetch-deps.

There must be a less fragile way of doing this. :thinking:

corngood replied:
@IvarWithoutBones this is now updated to move all the --runtime flags into the hooks, and remove the special logic here.

corngood commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -204,7 +198,7 @@ stdenvNoCC.mkDerivation (args // {
 
             # Since mktemp is used this will be empty if the script didnt succesfully complete
-            ! test -s "$depsFile" && rm -rf "$depsFile"
+            test -s "$depsFile" || rm -rf "$depsFile"
```
this was causing errors on exit from fetch-deps

corngood committed:
github-runner: use fake HOME in dependency fetcher

corngood committed:
treewide: update all dotnet lockfiles

corngood commented:
This now includes an update to all dotnet lockfiles (and a script to reproduce this).  Everything fetches without changes and builds on the last two commits.

mdarocha commented:
> This now includes an update to all dotnet lockfiles (and a script to reproduce this).  Everything fetches without changes and builds on the last two commits.

I'd remove this script:
1. It's not extensible - new packagers will probably not know to add their package to this script
2. It encourages global recreation of lockfiles that creates huge PR. We should instead be looking into ways to make changes to core functions independent of lockfiles, or at least backwards compatible.

mdarocha commented:
Also, you should name the PR and commit messages better, to indicate a tree wide change

corngood commented:
@mdarocha I removed the script for now.  I agree about it becoming out of date.  I'd love to find a way to do it based on package queries, but I haven't got to that yet.

The intention wasn't to encourage changes to the lock files.  It was to test changes to fetch-deps, nuget-to-nix, etc.  Some recent changes would have likely broken the update scripts for these packages.

I've updated the commit and PR messages.  I'm not particularly happy with the PR message, but I'm open to suggestions including splitting it up if that helps.

corngood commented:
sorry, I didn't realise this had picked up conflicts.  fixed

corngood commented:
Does anyone have any comments on this?  I'd like to get this in before more conflicts arise.

corngood commented:
@mdarocha rebased again.  anything I can do to help get this merged?

mdarocha commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -73,6 +71,7 @@
 , buildType ? "Release"
   # If set to true, builds the application as a self-contained - removing the runtime dependency on dotnet
 , selfContainedBuild ? false
+, useAppHost ? true
```
Is this flag used anywhere? Is it useful for it to be exposed?

GGG-KILLER replied:
I don't see any benefit for us to build .NET packages from source.

Are any other development environment's packages build by us? I don't see this with node ones at least (and Rust ones are always build from source but not directly by us).
What would the benefits of building it ourselves be, other than increasing maintenance cost even more?

corngood replied:
One of the great things about nixpkgs is being able to easily patch or reconfigure any package used on your system.

I just used newtonsoft-json as a test case because it's simple and ubiquitous.  I don't plan on proposing any major changes until it can be done with a reasonable maintenance burden.

Surely anything that can be built from source should be.  The fact that you don't see the benefit of this makes me worry that it's an uphill battle.  I'd love to discuss that but it's a little off-topic for here.

GGG-KILLER replied:
> Surely anything that can be built from source should be. The fact that you don't see the benefit of this makes me worry that it's an uphill battle. I'd love to discuss that but it's a little off-topic for here.

If you want, you can open an issue and we can discuss that there

corngood replied:
If you can't disable this, when you try to build a project using netcoreapp2.0 (e.g. newtonsoft-json 11, which is still commonly used), you get something like:

```
/nix/store/1iwq17ry9s985r1bxyflcca51cpaipvq-dotnet-sdk-5.0.408/sdk/5.0.408/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(138,5): error NETSDK1068: The framework-dependent application host requires a target framework of at least 'netcoreapp2.1'. [/build/source/Src/Newtonsoft.Json.Tests/Newtonsoft.Json.Tests.csproj]
```

mdarocha replied:
I guess that's a good enough reason to leave it, we currently support building library projects and referencing them

mdarocha commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -90,6 +89,7 @@ let
 
   inherit (callPackage ./hooks {
     inherit dotnet-sdk dotnet-test-sdk disabledTests nuget-source dotnet-runtime runtimeDeps buildType;
+    runtimeId = dotnetCorePackages.systemToDotnetRid stdenvNoCC.targetPlatform.system;
```
I would prefer for this value to be exposed in the top 'buildDotnetModule' as an argument, with the default value set as above. Then nixpkgs pkgs would use the default value, but you also wouldn't break some workflows that use this function outside nixpkgs (like ie. cross-building for windows by setting the runtime id)

corngood replied:
I made this change:

```diff
commit 18847999cad9611178d086b8d53877877a2cbd2f
Author: David McFarland <corngood@gmail.com>
Date:   Thu Dec 8 12:40:10 2022 -0400

    fixup! build-dotnet-module: restore for current runtime by default

diff --git a/pkgs/build-support/dotnet/build-dotnet-module/default.nix b/pkgs/build-support/dotnet/build-dotnet-module/default.nix
index e297bd3bbcd..273dbd35ff7 100644
--- a/pkgs/build-support/dotnet/build-dotnet-module/default.nix
+++ b/pkgs/build-support/dotnet/build-dotnet-module/default.nix
@@ -59,6 +59,9 @@
   # Libraries that need to be available at runtime should be passed through this.
   # These get wrapped into `LD_LIBRARY_PATH`.
 , runtimeDeps ? [ ]
+  # The dotnet runtime ID. If null, fetch-deps will gather dependencies for all
+  # platforms in meta.platforms which are supported by the sdk.
+, runtimeId ? null
 
   # Tests to disable. This gets passed to `dotnet test --filter "FullyQualifiedName!={}"`, to ensure compatibility with all frameworks.
   # See https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-test#filter-option-details for more details.
@@ -90,7 +93,10 @@ let
 
   inherit (callPackage ./hooks {
     inherit dotnet-sdk dotnet-test-sdk disabledTests nuget-source dotnet-runtime runtimeDeps buildType;
-    runtimeId = dotnetCorePackages.systemToDotnetRid stdenvNoCC.targetPlatform.system;
+    runtimeId =
+      if runtimeId != null
+      then [ runtimeId ]
+      else dotnetCorePackages.systemToDotnetRid stdenvNoCC.targetPlatform.system;
   }) dotnetConfigureHook dotnetBuildHook dotnetCheckHook dotnetInstallHook dotnetFixupHook;
 
   localDeps =
@@ -158,7 +164,10 @@ stdenvNoCC.mkDerivation (args // {
     fetch-deps =
       let
         flags = dotnetFlags ++ dotnetRestoreFlags;
-        runtimeIds = map (system: dotnetCorePackages.systemToDotnetRid system) platforms;
+        runtimeIds =
+          if runtimeId != null
+          then [ runtimeId ]
+          else map (system: dotnetCorePackages.systemToDotnetRid system) platforms;
       in
       writeShellScript "fetch-${pname}-deps" ''
         set -euo pipefail
```

I was able to fetch-deps and build a dotnet package on linux using `runtimeId = "win-x64"`.

One potential problem with this is that you can't specify multiple RIDs to restore in fetch-deps.  Is that a problem for you?

If you have any examples of how you use this, I could make some tests.

mdarocha replied:
Yeah, this fixes my use case and should be enough.

I don't see the problem with the `fetch-deps` logic, it should always restore for all platforms specified. Maybe include windows rids in `systemToDotnetRid`? I know nixpkgs has windows targets specified.

corngood replied:
I added RIDs for the two windows configs in doubles.nix.

mdarocha commented:
> @mdarocha rebased again.  anything I can do to help get this merged?

Well, I suggested some changes, since they break some workflows (including mine :))

GGG-KILLER commented:
> Well, I suggested some changes, since they break some workflows (including mine :))
> 
> Maybe ie. @GGG-KILLER or @winterqt can take a look at this pull request?

@mdarocha and @corngood I've checked the PR previously but don't have anything to add other than you already have

corngood committed:
dotnet: add windows RIDs to systemToDotnetRid

mdarocha commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -90,6 +93,10 @@ let
     inherit dotnet-sdk dotnet-test-sdk disabledTests nuget-source dotnet-runtime runtimeDeps buildType;
+    runtimeId =
+      if runtimeId != null
+      then [ runtimeId ]
```
why is it put into an array?

corngood replied:
Good catch, that was unintentional.  I added an assert for the type in hooks/default.nix.

mdarocha commented:
> @mdarocha any more thoughts on the open issues?  Could you resolve them if you're happy?

It all LGTM except one question I posted now.

I unfortunately don't have the rights to merge this, you have to find and tag somebody with merge access or post it in the "PRs ready for review" thread on discourse

corngood commented:
@AndersonTorres you merged my last dotnet related change.  Perhaps you could take a look?

AndersonTorres commented:
Given that it was approved by package maintainers, I will merge this.

chuangzhu commented on pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-install-hook.sh:
```diff
@@ -7,17 +7,20 @@ dotnetInstallHook() {
 
     if [ "${selfContainedBuild-}" ]; then
-        dotnetInstallFlags+=("--self-contained")
+        dotnetInstallFlags+=(--runtime "@runtimeId@" "--self-contained")
```
Why is this only for selfContainedBuild?

corngood replied:
Sorry for the late reply, but `--self-contained` without '--runtime' results in:

```
/nix/store/0jarb9qww0jg62lw6mlkjgs9fx3gxw98-dotnet-sdk-6.0.300/sdk/6.0.300/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(151,5): error NETSDK1031: It is not supported to build or publish a self-contained application without specifying a RuntimeIdentifier. You must either specify a RuntimeIdentifier or set SelfContained to false. [/home/davidm/tmp/tmp.csproj]
```

`--no-self-contained` works without a runtime id.  Presumably the rid only applies when making a self-contained executable.

chuangzhu replied:
If `--runtime` is omitted in a `--no-self-contained` build, native libraries for all platforms will be installed in `result/lib/<name>/runtimes`. There is actually an example using `--runtime` and `--self-contained false` together [in dotnet documentations](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-publish#examples).
